File: sample/src/main/java/me/xiaopan/sketchsample/widget/FindEmptyView.java
package me.xiaopan.sketchsample.widget;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Rect;
import android.util.AttributeSet;
import android.view.View;

import java.util.List;

import me.xiaopan.sketch.util.SketchUtils;

public class FindEmptyView extends View {
    private List<Rect> fullRectList;
    private List<Rect> emptyRectList;
    private Rect boundsRect;

    private Paint boundsRectPaint;
    private Paint fullRectPaint;
    private Paint emptyRectPaint;

    public FindEmptyView(Context context) {
        super(context);
        init(context);
    }

    public FindEmptyView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(context);
    }

    public FindEmptyView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(context);
    }

    private void init(Context context) {
        boundsRectPaint = new Paint();
        boundsRectPaint.setStyle(Paint.Style.STROKE);
        boundsRectPaint.setColor(Color.parseColor("#8800CD00"));
        boundsRectPaint.setStrokeWidth(SketchUtils.dp2px(context, 1));

        fullRectPaint = new Paint();
        fullRectPaint.setColor(Color.parseColor("#88FF0000"));
        fullRectPaint.setStrokeWidth(SketchUtils.dp2px(context, 1));
        fullRectPaint.setStyle(Paint.Style.STROKE);

        emptyRectPaint = new Paint();
        emptyRectPaint.setColor(Color.parseColor("#880000CD"));
        emptyRectPaint.setStrokeWidth(SketchUtils.dp2px(context, 1));
        emptyRectPaint.setStyle(Paint.Style.STROKE);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);

        if (fullRectList != null) {
            for (Rect fullRect : fullRectList) {
                if (!fullRect.isEmpty()) {
                    canvas.drawRect(fullRect.left * 3 + 1, fullRect.top * 3 + 1, fullRect.right * 3 - 1, fullRect.bottom * 3 - 1, fullRectPaint);
                }
            }
        }

        if (emptyRectList != null) {
            for (Rect emptyRect : emptyRectList) {
                if (!emptyRect.isEmpty()) {
                    canvas.drawRect(emptyRect.left * 3 + 1, emptyRect.top * 3 + 1, emptyRect.right * 3 - 1, emptyRect.bottom * 3 - 1, emptyRectPaint);
                }
            }
        }

        if (boundsRect != null && !boundsRect.isEmpty()) {
            canvas.drawRect(boundsRect.left * 3, boundsRect.top * 3, boundsRect.right * 3, boundsRect.bottom * 3, boundsRectPaint);
        }
    }

    public void setBoundsRect(Rect boundsRect) {
        this.boundsRect = boundsRect;
    }

    public void setEmptyRectList(List<Rect> emptyRectList) {
        this.emptyRectList = emptyRectList;
    }

    public void setFullRectList(List<Rect> fullRectList) {
        this.fullRectList = fullRectList;
    }
}

File: sample/src/main/java/me/xiaopan/sketchsample/widget/FindEmptyView.kt
package me.xiaopan.sketchsample.widget

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Rect
import android.util.AttributeSet
import android.view.View

import me.xiaopan.sketch.util.SketchUtils

class FindEmptyView : View {
    private var fullRectList: List<Rect>? = null
    private var emptyRectList: List<Rect>? = null
    private var boundsRect: Rect? = null

    private val boundsRectPaint: Paint = Paint()
    private val fullRectPaint: Paint = Paint()
    private val emptyRectPaint: Paint = Paint()

    constructor(context: Context) : super(context) {
        init(context)
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        init(context)
    }

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        init(context)
    }

    private fun init(context: Context) {
        boundsRectPaint.style = Paint.Style.STROKE
        boundsRectPaint.color = Color.parseColor("#8800CD00")
        boundsRectPaint.strokeWidth = SketchUtils.dp2px(context, 1).toFloat()

        fullRectPaint.color = Color.parseColor("#88FF0000")
        fullRectPaint.strokeWidth = SketchUtils.dp2px(context, 1).toFloat()
        fullRectPaint.style = Paint.Style.STROKE

        emptyRectPaint.color = Color.parseColor("#880000CD")
        emptyRectPaint.strokeWidth = SketchUtils.dp2px(context, 1).toFloat()
        emptyRectPaint.style = Paint.Style.STROKE
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)

        if (fullRectList != null) {
            for (fullRect in fullRectList!!) {
                if (!fullRect.isEmpty) {
                    canvas.drawRect(
                            (fullRect.left * 3 + 1).toFloat(),
                            (fullRect.top * 3 + 1).toFloat(),
                            (fullRect.right * 3 - 1).toFloat(),
                            (fullRect.bottom * 3 - 1).toFloat(),
                            fullRectPaint)
                }
            }
        }

        if (emptyRectList != null) {
            for (emptyRect in emptyRectList!!) {
                if (!emptyRect.isEmpty) {
                    canvas.drawRect(
                            (emptyRect.left * 3 + 1).toFloat(),
                            (emptyRect.top * 3 + 1).toFloat(),
                            (emptyRect.right * 3 - 1).toFloat(),
                            (emptyRect.bottom * 3 - 1).toFloat(),
                            emptyRectPaint)
                }
            }
        }

        if (boundsRect != null && !boundsRect!!.isEmpty) {
            canvas.drawRect((boundsRect!!.left * 3).toFloat(), (boundsRect!!.top * 3).toFloat(), (boundsRect!!.right * 3).toFloat(), (boundsRect!!.bottom * 3).toFloat(), boundsRectPaint)
        }
    }

    fun setBoundsRect(boundsRect: Rect) {
        this.boundsRect = boundsRect
    }

    fun setEmptyRectList(emptyRectList: List<Rect>?) {
        this.emptyRectList = emptyRectList
    }

    fun setFullRectList(fullRectList: List<Rect>) {
        this.fullRectList = fullRectList
    }
}
